TypeError in tsc when destructuring assignment from ES6 object literal syntax

**TypeScript Version:**  3.9.0-dev.20200424

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** UnhandledPromiseRejectionWarning

**Code**

*test.js*
```js
function fn(obj, k, v) {
  // this breaks
  const { x } = { ...obj, [k]: v }
  // but this works:
  const x = { ...obj, [k]: v }
}
```

Run `tsc --allowJs test.js`

**Expected behavior:**

The output should be the following:
```
error TS5055: Cannot write file 'test.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.


Found 1 error.
```

**Actual behavior:**

```
/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:82820
                throw e;
                ^

TypeError: Cannot read property 'name' of undefined
    at getObjectLiteralIndexInfo (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:46718:77)
    at createObjectLiteralType (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:46887:67)
    at checkObjectLiteral (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:46876:52)
    at checkExpressionWorker (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:51490:28)
    at checkExpression (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:51419:38)
    at checkExpressionCached (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:51145:38)
    at checkDeclarationInitializer (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:51158:114)
    at getTypeForVariableLikeDeclaration (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:34194:74)
    at getTypeForBindingElementParent (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:33966:61)
    at checkVariableLikeDeclaration (/Users/naman/dev/go/src/github.com/kiteco/kiteco/sidebar/node_modules/typescript/lib/tsc.js:53361:34)
```

**Playground Link:** I can't figure out how to make this happen in the playground

**Related Issues:** nope

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -23470,8 +23470,8 @@ namespace ts {
 
         function getObjectLiteralIndexInfo(node: ObjectLiteralExpression, offset: number, properties: Symbol[], kind: IndexKind): IndexInfo {
             const propTypes: Type[] = [];
-            for (let i = 0; i < properties.length; i++) {
-                if (kind === IndexKind.String || isNumericName(node.properties[i + offset].name!)) {
+            for (let i = offset; i < properties.length; i++) {
+                if (kind === IndexKind.String || isNumericName(node.properties[i].name!)) {
                     propTypes.push(getTypeOfSymbol(properties[i]));
                 }
             }

--- a/tests/baselines/reference/checkDestructuringShorthandAssigment2.errors.txt
+++ b/tests/baselines/reference/checkDestructuringShorthandAssigment2.errors.txt
@@ -1,0 +1,14 @@
+tests/cases/compiler/checkDestructuringShorthandAssigment2.ts(4,7): error TS2525: Initializer provides no value for this binding element and the binding element has no default value.
+tests/cases/compiler/checkDestructuringShorthandAssigment2.ts(4,27): error TS2353: Object literal may only specify known properties, and '[k]' does not exist in type '{ x: any; }'.
+
+
+==== tests/cases/compiler/checkDestructuringShorthandAssigment2.ts (2 errors) ====
+    // GH #38175 -- should not crash while checking
+    
+    let o: any, k: any;
+    let { x } = { x: 1, ...o, [k]: 1 };
+          ~
+!!! error TS2525: Initializer provides no value for this binding element and the binding element has no default value.
+                              ~~~
+!!! error TS2353: Object literal may only specify known properties, and '[k]' does not exist in type '{ x: any; }'.
+    

--- a/tests/baselines/reference/checkDestructuringShorthandAssigment2.symbols
+++ b/tests/baselines/reference/checkDestructuringShorthandAssigment2.symbols
@@ -1,0 +1,14 @@
+=== tests/cases/compiler/checkDestructuringShorthandAssigment2.ts ===
+// GH #38175 -- should not crash while checking
+
+let o: any, k: any;
+>o : Symbol(o, Decl(checkDestructuringShorthandAssigment2.ts, 2, 3))
+>k : Symbol(k, Decl(checkDestructuringShorthandAssigment2.ts, 2, 11))
+
+let { x } = { x: 1, ...o, [k]: 1 };
+>x : Symbol(x, Decl(checkDestructuringShorthandAssigment2.ts, 3, 5))
+>x : Symbol(x, Decl(checkDestructuringShorthandAssigment2.ts, 3, 13))
+>o : Symbol(o, Decl(checkDestructuringShorthandAssigment2.ts, 2, 3))
+>[k] : Symbol([k], Decl(checkDestructuringShorthandAssigment2.ts, 3, 25))
+>k : Symbol(k, Decl(checkDestructuringShorthandAssigment2.ts, 2, 11))
+

--- a/tests/baselines/reference/checkDestructuringShorthandAssigment2.types
+++ b/tests/baselines/reference/checkDestructuringShorthandAssigment2.types
@@ -1,0 +1,17 @@
+=== tests/cases/compiler/checkDestructuringShorthandAssigment2.ts ===
+// GH #38175 -- should not crash while checking
+
+let o: any, k: any;
+>o : any
+>k : any
+
+let { x } = { x: 1, ...o, [k]: 1 };
+>x : any
+>{ x: 1, ...o, [k]: 1 } : any
+>x : number
+>1 : 1
+>o : any
+>[k] : number
+>k : any
+>1 : 1
+

--- a/tests/cases/compiler/checkDestructuringShorthandAssigment2.ts
+++ b/tests/cases/compiler/checkDestructuringShorthandAssigment2.ts
@@ -1,0 +1,6 @@
+// @noEmit: true
+
+// GH #38175 -- should not crash while checking
+
+let o: any, k: any;
+let { x } = { x: 1, ...o, [k]: 1 };

